File: src/retar/solver.ts
import type { SystemRole, Seat } from '../types/index.ts'
import { dumpSolveResult } from './dump.ts'
import {
  Possibilities,
  RoleSignatureBits,
  popCount,
  combinationWithReplacementBit,
  bitIndicesFromMask,
  RoleBitIndex,
  ROLE_COUNT,
} from './possibilities.ts'
import type { RolePossibility } from './possibilities.ts'
import { rolesBySeerResult } from './role-sets.ts'
import { systemRoles } from '../types/index.ts'

// 「seerResult='wolf' な役職」 全集合 (現状は werewolf 1 種、 将来複数化に対応).
// HAMSTER_BITS / FOX_SIGNATURES と同パターン.
const WOLF_BITS: number[] = rolesBySeerResult('wolf').map(r => RoleBitIndex[r])
const WOLF_SIGNATURES: number[] = WOLF_BITS.map(bit => 1 << bit)

// 狐陣営勝利の生存カウント対象 = passive:fox-win-counter trait を持つ全役職 (妖狐 + 子狐 + 将来追加).
// passive:die-when-divined (= 妖狐のみ) で数えると子狐生存パスが棄却される.
const HAMSTER_BITS: number[] = Array.from(systemRoles.entries())
  .filter(([, meta]) => meta.traits.some(t => t.kind === 'passive' && t.sub === 'fox-win-counter'))
  .map(([role]) => RoleBitIndex[role])
const FOX_SIGNATURES: number[] = HAMSTER_BITS.map(bit => 1 << bit)

/*
 * Solver configuration — immutable across recursion.
 * These values define the search space and are threaded through
 * the recursive backtracking without changing.
 */
type SolverConfig = {
  conclusion: Possibilities
  items: ([RolePossibility, number[]] | 'check')[]
  wolvesRange: [min: number, max: number]
  hamstersRange: [min: number, max: number]
}

/*
 * Role assignment backtracking solver.
 *
 * The `items` array is structured as:
 *   [ ...fixed(dead), ...fixed(alive), ...survivors, "check", ...dead ]
 *
 * - Fixed seats (popCount=1) come first regardless of alive/dead status.
 * - Unfixed survivors come next — all valid role combinations are enumerated.
 * - "check" sentinel validates wolf/hamster survival counts at the boundary.
 * - Unfixed dead seats come last — only satisfiability is checked (not all solutions),
 *   because enumerating all dead combinations would explode combinatorially.
 *
 * The solver writes valid role possibilities into `config.conclusion` via bitwise OR,
 * accumulating the union of all valid assignments.
 *
 * roleCount: Uint8Array[ROLE_COUNT] indexed by bit position (0=villager, ..., 10=immoralist)
 */
function backtrackForRoleAssignment(
  config: SolverConfig,
  roleCount: Uint8Array,
  index: number,
  selectedWolves: number,
  selectedHamsters: number,
  path: [number[], Uint8Array][],
  all: boolean,
): boolean {
  const item = config.items[index]
  if (item === 'check') {
    if (config.wolvesRange[1] < selectedWolves) return false
    if (selectedWolves < config.wolvesRange[0]) return false
    if (config.hamstersRange[1] < selectedHamsters) return false
    if (selectedHamsters < config.hamstersRange[0]) return false
    const res = backtrackForRoleAssignment(
      config,
      new Uint8Array(roleCount),
      index + 1,
      selectedWolves,
      selectedHamsters,
      [],
      false,
    )
    if (!res) return false
    // Build bitmask of roles with remaining count > 0
    let filterForDeads = 0
    for (let i = 0; i < ROLE_COUNT; i++) {
      if (roleCount[i] > 0) filterForDeads |= (1 << i)
    }
    // Collect dead seat entries with filtered possibilities
    const deadEntries: [RolePossibility, number[]][] = []
    for (let i = index + 1; i < config.items.length; i++) {
      const item = config.items[i]
      if (item === 'check') continue
      const [possibility, seats] = item
      deadEntries.push([possibility & filterForDeads, seats])
    }
    // Propagate constraints among dead seats:
    // If a group has only one possible role and that role's remaining count
    // equals the group size, remove that role from all other groups
    const deadRoleCount = new Uint8Array(roleCount)
    let changed = true
    while (changed) {
      changed = false
      for (const entry of deadEntries) {
        if (popCount(entry[0]) === 1) {
          const bitIdx = 31 - Math.clz32(entry[0])
          if (deadRoleCount[bitIdx] > 0 && deadRoleCount[bitIdx] === entry[1].length) {
            deadRoleCount[bitIdx] = 0
            for (const other of deadEntries) {
              if (other === entry) continue
              const before = other[0]
              other[0] = before & ~entry[0]
              if (other[0] !== before) changed = true
            }
          }
        }
      }
      // Naked subset: groups whose possibilities ⊆ mask consume exactly those roles
      const checked = new Set<number>()
      for (const entry of deadEntries) {
        const mask = entry[0]
        if (mask === 0 || checked.has(mask)) continue
        checked.add(mask)
        let roleSum = 0
        for (let i = 0; i < ROLE_COUNT; i++) {
          if (mask & (1 << i)) roleSum += deadRoleCount[i]
        }
        if (roleSum === 0) continue
        let seatSum = 0
        for (const e of deadEntries) {
          if (e[0] !== 0 && (e[0] & mask) === e[0]) seatSum += e[1].length
        }
        if (seatSum === roleSum) {
          for (const e of deadEntries) {
            if ((e[0] & mask) === e[0]) continue
            const before = e[0]
            e[0] = before & ~mask
            if (e[0] !== before) changed = true
          }
        }
      }
      // Also check: if a role's remaining count is fully consumed by
      // groups that must include it, remove it from other groups
      for (let bitIdx = 0; bitIdx < ROLE_COUNT; bitIdx++) {
        if (deadRoleCount[bitIdx] <= 0) continue
        const roleBit = 1 << bitIdx
        let totalSeats = 0
        const mustHaveGroups: typeof deadEntries = []
        for (const entry of deadEntries) {
          if (entry[0] & roleBit) {
            totalSeats += entry[1].length
            mustHaveGroups.push(entry)
          }
        }
        if (totalSeats === deadRoleCount[bitIdx]) {
          for (const entry of mustHaveGroups) {
            if (entry[0] !== roleBit) {
              entry[0] = roleBit
              changed = true
            }
          }
        }
      }
    }
    for (const [possibility, seats] of deadEntries) {
      for (const seat of seats) {
        config.conclusion.possibilities[seat] |= possibility
      }
    }
    for (const p of path) {
      const [seats, counts] = p
      for (let i = 0; i < ROLE_COUNT; i++) {
        if (counts[i] === 0) continue
        const bit = 1 << i
        for (const seat of seats) {
          config.conclusion.possibilities[seat] |= bit
        }
      }
    }
    return true
  }
  else if (index === config.items.length) {
    return true
  }
  else if (index === config.items.length - 1) {
    // Last element: check remaining role counts match seat count
    let count = 0
    let sub = 0
    for (let i = 0; i < ROLE_COUNT; i++) {
      if (roleCount[i] > 0) {
        count += roleCount[i]
        sub |= (1 << i)
      }
    }
    const [last, seats] = item as [RolePossibility, number[]]
    if (seats.length !== count) {
      return false
    }
    if ((last & sub) !== sub) {
      return false
    }
    return true
  }

  const [set, seats] = item as [RolePossibility, number[]]
  const indices = bitIndicesFromMask(set)
  let oneOK = false
  for (const v of combinationWithReplacementBit(indices, seats.length, roleCount)) {
    let ok = true
    for (const idx of indices) {
      if (v[idx] === 0) continue
      if (roleCount[idx] < v[idx]) ok = false
      roleCount[idx] -= v[idx]
    }
    path.push([seats, new Uint8Array(v)])
    if (ok) {
      let addedHamsters = 0
      for (const bit of HAMSTER_BITS) addedHamsters += v[bit]
      let addedWolves = 0
      for (const bit of WOLF_BITS) addedWolves += v[bit]
      const res = backtrackForRoleAssignment(
        config,
        roleCount,
        index + 1,
        selectedWolves + addedWolves,
        selectedHamsters + addedHamsters,
        path,
        all,
      )
      if (res && !all) return true
      if (res) oneOK = true
    }
    for (const idx of indices) {
      if (v[idx] === 0) continue
      roleCount[idx] += v[idx]
    }
    path.pop()
  }
  return oneOK
}

/*
 * Solve possible role assignments for all seats.
 *
 * Groups seats by their possibility bitmask (seats with identical possibilities
 * are tested together for efficiency). Fixed seats are processed first, then
 * survivors (full enumeration), then dead seats (satisfiability only).
 *
 * Returns a Possibilities object with the union of all valid assignments,
 * or undefined if no valid assignment exists.
 */
export function solvePossibilities(
  source: Possibilities,
  survivors: Map<Seat, boolean>,
  minSurvivingWolves: number,
  maxSurvivingWolves: number,
  minSurvivingHamsters: number,
  maxSurvivingHamsters: number,
  setup: Map<SystemRole, number>,
): Possibilities | undefined {
  // Group seats by possibility bitmask, separated by survival status
  const survivorsMap: Map<number, number[]> = new Map()
  const deadMap: Map<number, number[]> = new Map()
  const fixedMap: Map<number, number[]> = new Map()
  let fixedDiedWolves: number = 0
  let fixedDiedHamsters: number = 0
  for (let i = 1; i < source.possibilities.length; i++) {
    const possibility = source.possibilities[i]
    const count = popCount(possibility)
    if (count === 0) return undefined
    if (count === 1) {
      if (!fixedMap.has(possibility)) {
        fixedMap.set(possibility, [])
      }
      fixedMap.get(possibility)!.push(i)
      if (WOLF_SIGNATURES.includes(possibility) && !survivors.get(i)) {
        fixedDiedWolves++
      }
      if (!survivors.get(i) && FOX_SIGNATURES.includes(possibility)) {
        fixedDiedHamsters++
      }
      continue
    }
    if (survivors.get(i)) {
      if (!survivorsMap.has(possibility)) {
        survivorsMap.set(possibility, [])
      }
      survivorsMap.get(possibility)!.push(i)
    } else {
      if (!deadMap.has(possibility)) {
        deadMap.set(possibility, [])
      }
      deadMap.get(possibility)!.push(i)
    }
  }
  const items: ([number, number[]] | 'check')[]
    = Array.from(deadMap.size
      ? [...fixedMap.entries(), ...survivorsMap.entries(), 'check' as const, ...deadMap.entries()]
      : [...fixedMap.entries(), ...survivorsMap.entries(), 'check' as const]
    )

  const config: SolverConfig = {
    conclusion: Possibilities.empty(setup),
    items,
    wolvesRange: [fixedDiedWolves + minSurvivingWolves, fixedDiedWolves + maxSurvivingWolves],
    hamstersRange: [fixedDiedHamsters + minSurvivingHamsters, fixedDiedHamsters + maxSurvivingHamsters],
  }

  // Build initial roleCount as Uint8Array from setup
  const roleCount = new Uint8Array(ROLE_COUNT)
  for (const [role, count] of setup) {
    roleCount[RoleBitIndex[role]] = count
  }

  const res = backtrackForRoleAssignment(
    config,
    roleCount,
    0, 0, 0, [], true,
  )
  if (!res) {
    dumpSolveResult(undefined)
    return undefined
  }
  dumpSolveResult(config.conclusion)
  return config.conclusion
}

File: src/retar-rs/src/solver.rs
use crate::types::{SystemRole, Seat};
use crate::possibilities::{
    Possibilities, ROLE_COUNT, pop_count, bit_indices_from_mask,
    combination_with_replacement_bit,
};
use std::collections::BTreeMap;

const WOLF_BIT: usize = SystemRole::Werewolf.bit_index_const() as usize;
const HAMSTER_BIT: usize = SystemRole::Werehamster.bit_index_const() as usize;

/// Item in the solver's work list.
#[derive(Debug, Clone)]
pub enum SolverItem {
    /// A group of seats sharing the same possibility bitmask.
    Group(u16, Vec<Seat>),
    /// Sentinel separating survivors from dead seats.
    Check,
}

/// Solver configuration — immutable across recursion.
struct SolverConfig {
    conclusion: Possibilities,
    items: Vec<SolverItem>,
    wolves_range: (u32, u32),   // (min, max) total wolves
    hamsters_range: (u32, u32), // (min, max) total hamsters
}

/// A path entry: (seats, role counts assigned to those seats).
type PathEntry = (Vec<Seat>, [u8; ROLE_COUNT]);

fn backtrack_for_role_assignment(
    config: &mut SolverConfig,
    role_count: &mut [u8; ROLE_COUNT],
    index: usize,
    selected_wolves: u32,
    selected_hamsters: u32,
    path: &mut Vec<PathEntry>,
    all: bool,
) -> bool {
    if index >= config.items.len() {
        return true;
    }

    match &config.items[index] {
        SolverItem::Check => {
            // Validate wolf/hamster counts at the survivor/dead boundary
            if config.wolves_range.1 < selected_wolves {
                return false;
            }
            if selected_wolves < config.wolves_range.0 {
                return false;
            }
            if config.hamsters_range.1 < selected_hamsters {
                return false;
            }
            if selected_hamsters < config.hamsters_range.0 {
                return false;
            }

            let mut role_count_copy = *role_count;
            let res = backtrack_for_role_assignment(
                config,
                &mut role_count_copy,
                index + 1,
                selected_wolves,
                selected_hamsters,
                &mut Vec::new(),
                false,
            );
            if !res {
                return false;
            }

            // Build bitmask of roles with remaining count > 0
            let mut filter_for_deads: u16 = 0;
            for i in 0..ROLE_COUNT {
                if role_count[i] > 0 {
                    filter_for_deads |= 1u16 << i;
                }
            }

            // Collect dead seat entries with filtered possibilities
            let mut dead_entries: Vec<(u16, Vec<Seat>)> = Vec::new();
            for i in (index + 1)..config.items.len() {
                if let SolverItem::Group(possibility, seats) = &config.items[i] {
                    dead_entries.push((possibility & filter_for_deads, seats.clone()));
                }
            }

            // Propagate constraints among dead seats
            let mut dead_role_count = *role_count;
            let mut changed = true;
            while changed {
                changed = false;

                // Naked singles: if a group has exactly one role and that role's count matches group size
                for idx in 0..dead_entries.len() {
                    let (poss, seats) = &dead_entries[idx];
                    if pop_count(*poss) == 1 {
                        let bit_idx = poss.trailing_zeros() as usize;
                        if dead_role_count[bit_idx] > 0
                            && dead_role_count[bit_idx] == seats.len() as u8
                        {
                            dead_role_count[bit_idx] = 0;
                            let mask = *poss;
                            for other_idx in 0..dead_entries.len() {
                                if other_idx == idx {
                                    continue;
                                }
                                let before = dead_entries[other_idx].0;
                                dead_entries[other_idx].0 = before & !mask;
                                if dead_entries[other_idx].0 != before {
                                    changed = true;
                                }
                            }
                        }
                    }
                }

                // Naked subset: groups whose possibilities ⊆ mask consume exactly those roles
                let mut checked = std::collections::BTreeSet::new();
                for idx in 0..dead_entries.len() {
                    let mask = dead_entries[idx].0;
                    if mask == 0 || checked.contains(&mask) {
                        continue;
                    }
                    checked.insert(mask);
                    let mut role_sum: u32 = 0;
                    for i in 0..ROLE_COUNT {
                        if mask & (1u16 << i) != 0 {
                            role_sum += dead_role_count[i] as u32;
                        }
                    }
                    if role_sum == 0 {
                        continue;
                    }
                    let mut seat_sum: u32 = 0;
                    for e in &dead_entries {
                        if e.0 != 0 && (e.0 & mask) == e.0 {
                            seat_sum += e.1.len() as u32;
                        }
                    }
                    if seat_sum == role_sum {
                        for e in &mut dead_entries {
                            if (e.0 & mask) == e.0 {
                                continue;
                            }
                            let before = e.0;
                            e.0 = before & !mask;
                            if e.0 != before {
                                changed = true;
                            }
                        }
                    }
                }

                // Role exhaustion: if a role's remaining count equals the total seats that need it
                for bit_idx in 0..ROLE_COUNT {
                    if dead_role_count[bit_idx] == 0 {
                        continue;
                    }
                    let role_bit = 1u16 << bit_idx;
                    let mut total_seats: u32 = 0;
                    let mut must_have_indices: Vec<usize> = Vec::new();
                    for (i, entry) in dead_entries.iter().enumerate() {
                        if entry.0 & role_bit != 0 {
                            total_seats += entry.1.len() as u32;
                            must_have_indices.push(i);
                        }
                    }
                    if total_seats == dead_role_count[bit_idx] as u32 {
                        for &i in &must_have_indices {
                            if dead_entries[i].0 != role_bit {
                                dead_entries[i].0 = role_bit;
                                changed = true;
                            }
                        }
                    }
                }
            }

            // Write dead seat results into conclusion
            for (possibility, seats) in &dead_entries {
                for &seat in seats {
                    config.conclusion.possibilities[seat as usize] |= possibility;
                }
            }

            // Write survivor path results into conclusion
            for (seats, counts) in path.iter() {
                for i in 0..ROLE_COUNT {
                    if counts[i] == 0 {
                        continue;
                    }
                    let bit = 1u16 << i;
                    for &seat in seats {
                        config.conclusion.possibilities[seat as usize] |= bit;
                    }
                }
            }

            true
        }
        SolverItem::Group(_, _) => {
            // Clone to avoid borrow issues
            let (set, seats) = match &config.items[index] {
                SolverItem::Group(s, seats) => (*s, seats.clone()),
                _ => unreachable!(),
            };

            // Last element optimization: check remaining role counts match seat count
            if index == config.items.len() - 1 {
                let mut count: u32 = 0;
                let mut sub: u16 = 0;
                for i in 0..ROLE_COUNT {
                    if role_count[i] > 0 {
                        count += role_count[i] as u32;
                        sub |= 1u16 << i;
                    }
                }
                if seats.len() as u32 != count {
                    return false;
                }
                if (set & sub) != sub {
                    return false;
                }
                return true;
            }

            let indices = bit_indices_from_mask(set);
            let mut one_ok = false;

            // Collect all combinations first (snapshot of limits), then process
            let limits_snapshot = *role_count;
            let mut combos: Vec<[u8; ROLE_COUNT]> = Vec::new();
            combination_with_replacement_bit(
                &indices,
                seats.len() as u32,
                &limits_snapshot,
                &mut |v| {
                    combos.push(*v);
                    true
                },
            );

            for v in &combos {
                let mut ok = true;
                for &idx in &indices {
                    let idx = idx as usize;
                    if v[idx] == 0 {
                        continue;
                    }
                    if role_count[idx] < v[idx] {
                        ok = false;
                    }
                    role_count[idx] -= v[idx];
                }
                path.push((seats.clone(), *v));
                if ok {
                    let res = backtrack_for_role_assignment(
                        config,
                        role_count,
                        index + 1,
                        selected_wolves + v[WOLF_BIT] as u32,
                        selected_hamsters + v[HAMSTER_BIT] as u32,
                        path,
                        all,
                    );
                    if res && !all {
                        for &idx in &indices {
                            let idx = idx as usize;
                            if v[idx] == 0 {
                                continue;
                            }
                            role_count[idx] += v[idx];
                        }
                        path.pop();
                        return true;
                    }
                    if res {
                        one_ok = true;
                    }
                }
                for &idx in &indices {
                    let idx = idx as usize;
                    if v[idx] == 0 {
                        continue;
                    }
                    role_count[idx] += v[idx];
                }
                path.pop();
            }
            one_ok
        }
    }
}

/// Solve possible role assignments for all seats.
///
/// Groups seats by their possibility bitmask (seats with identical possibilities
/// are tested together for efficiency). Fixed seats are processed first, then
/// survivors (full enumeration), then dead seats (satisfiability only).
///
/// Returns a Possibilities object with the union of all valid assignments,
/// or None if no valid assignment exists.
pub fn solve_possibilities(
    source: &Possibilities,
    survivors: &BTreeMap<Seat, bool>,
    min_surviving_wolves: u32,
    max_surviving_wolves: u32,
    min_surviving_hamsters: u32,
    max_surviving_hamsters: u32,
    setup: &BTreeMap<SystemRole, u32>,
) -> Option<Possibilities> {
    // Group seats by possibility bitmask, separated by survival status
    let mut survivors_map: BTreeMap<u16, Vec<Seat>> = BTreeMap::new();
    let mut dead_map: BTreeMap<u16, Vec<Seat>> = BTreeMap::new();
    let mut fixed_map: BTreeMap<u16, Vec<Seat>> = BTreeMap::new();
    let mut fixed_died_wolves: u32 = 0;
    let mut fixed_died_hamsters: u32 = 0;

    for i in 1..source.possibilities.len() {
        let possibility = source.possibilities[i];
        let count = pop_count(possibility);
        if count == 0 {
            return None;
        }
        let seat = i as Seat;
        if count == 1 {
            fixed_map.entry(possibility).or_default().push(seat);
            if possibility == SystemRole::Werewolf.bit()
                && !survivors.get(&seat).copied().unwrap_or(false)
            {
                fixed_died_wolves += 1;
            }
            if possibility == SystemRole::Werehamster.bit()
                && !survivors.get(&seat).copied().unwrap_or(false)
            {
                fixed_died_hamsters += 1;
            }
            continue;
        }
        if survivors.get(&seat).copied().unwrap_or(false) {
            survivors_map.entry(possibility).or_default().push(seat);
        } else {
            dead_map.entry(possibility).or_default().push(seat);
        }
    }

    // Build items array: fixed, survivors, check, dead
    let mut items: Vec<SolverItem> = Vec::new();
    for (poss, seats) in &fixed_map {
        items.push(SolverItem::Group(*poss, seats.clone()));
    }
    for (poss, seats) in &survivors_map {
        items.push(SolverItem::Group(*poss, seats.clone()));
    }
    items.push(SolverItem::Check);
    if !dead_map.is_empty() {
        for (poss, seats) in &dead_map {
            items.push(SolverItem::Group(*poss, seats.clone()));
        }
    }

    let mut config = SolverConfig {
        conclusion: Possibilities::empty(setup),
        items,
        wolves_range: (
            fixed_died_wolves.saturating_add(min_surviving_wolves),
            fixed_died_wolves.saturating_add(max_surviving_wolves),
        ),
        hamsters_range: (
            fixed_died_hamsters.saturating_add(min_surviving_hamsters),
            fixed_died_hamsters.saturating_add(max_surviving_hamsters),
        ),
    };

    // Build initial role count
    let mut role_count = [0u8; ROLE_COUNT];
    for (&role, &count) in setup {
        role_count[role.bit_index() as usize] = count as u8;
    }

    let res = backtrack_for_role_assignment(
        &mut config,
        &mut role_count,
        0,
        0,
        0,
        &mut Vec::new(),
        true,
    );

    if !res {
        #[cfg(feature = "dump")] crate::dump::solve_result(None);
        return None;
    }
    #[cfg(feature = "dump")] crate::dump::solve_result(Some(&config.conclusion));
    Some(config.conclusion)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_setup(pairs: &[(SystemRole, u32)]) -> BTreeMap<SystemRole, u32> {
        pairs.iter().cloned().collect()
    }

    #[test]
    fn solve_simple_fully_fixed() {
        // 4 seats, 4 unique roles — should trivially solve
        let setup = make_setup(&[
            (SystemRole::Seer, 1),
            (SystemRole::Bodyguard, 1),
            (SystemRole::Werewolf, 1),
            (SystemRole::Possessed, 1),
        ]);
        let mut p = Possibilities::from_setup(&setup);
        p.fix_role(1, SystemRole::Seer);
        p.fix_role(2, SystemRole::Bodyguard);
        p.fix_role(3, SystemRole::Werewolf);
        p.fix_role(4, SystemRole::Possessed);

        let survivors: BTreeMap<Seat, bool> =
            [(1, true), (2, true), (3, true), (4, true)].into_iter().collect();

        let result = solve_possibilities(&p, &survivors, 0, 2, 0, 0, &setup);
        assert!(result.is_some());
    }

    #[test]
    fn solve_with_unfixed_seats() {
        // 3 seats, seer/werewolf/villager — seat 1 dead (night kill)
        let setup = make_setup(&[
            (SystemRole::Seer, 1),
            (SystemRole::Werewolf, 1),
            (SystemRole::Villager, 1),
        ]);
        let p = Possibilities::from_setup(&setup);

        // seat 1 dead, seats 2,3 alive
        let survivors: BTreeMap<Seat, bool> =
            [(1, false), (2, true), (3, true)].into_iter().collect();

        let result = solve_possibilities(&p, &survivors, 1, 1, 0, 0, &setup);
        assert!(result.is_some());
        let conclusion = result.unwrap();

        // Seat 1 is dead, so it can be seer or villager (not wolf since wolf must be alive)
        // Actually, wolf must survive (min=max=1 surviving wolves), so seat 1 can't be wolf
        // Seats 2,3 must include 1 wolf
        assert!(conclusion.has_role(2, SystemRole::Werewolf) || conclusion.has_role(3, SystemRole::Werewolf));
    }

    #[test]
    fn solve_returns_none_on_impossible() {
        // 2 seats, 2 wolves, but require 0 surviving wolves with all alive
        let setup = make_setup(&[(SystemRole::Werewolf, 2)]);
        let p = Possibilities::from_setup(&setup);
        let survivors: BTreeMap<Seat, bool> =
            [(1, true), (2, true)].into_iter().collect();

        // min/max surviving wolves = 0, but both seats are alive and must be wolves
        let result = solve_possibilities(&p, &survivors, 0, 0, 0, 0, &setup);
        assert!(result.is_none());
    }
}
